content/blog/running-self-hosted-agents-in-azure-devops-using-an-azure-ad-access-token.md

```diff
@@ -0,0 +1,126 @@
+These are pretty safe bets, as the best way to run the above is to just run it in an Azure Pipeline:
+
+```yaml
+---
+trigger:
+  - main
+
+pool:
+  vmImage: ubuntu-latest
+
+steps:
+- task: AzureCLI@2
+  inputs:
+    azureSubscription: YOUR_SERVICE_CONNECTION_HERE
+    scriptType: bash
+    scriptLocation: scriptPath
+    scriptPath: ./script.sh
+```
+
+It's easiest to set up the Service Connection (that will be passed to the `azureSubscription` input above) manually. Navigate to the project settings inside Azure Devops, create a new service connection (Azure Resource Manager -> Service Principal (Manual) and then fill in the ClientID, TenantID and either the Client Secret or the Client cert depending on what you set up earlier).
+
+Right now, this'll output an empty array, as we've not added any agents to this new agent pool yet (or maybe you did, I don't wanna make too many assumptions).
+
+So what's next? Well, we retrieve an Azure AD access token, and pass that along in place of the PAT, and we're set!
+
+One point of note, the Microsoft documentation (at this time of writing, I've sent them a note) is a little inconsistent as how to actually do this. At first, it refers to this [article](https://learn.microsoft.com/en-us/azure/active-directory/develop/v2-oauth2-client-creds-grant-flow#get-a-token) which led me to think it was as easy as doing a straightforward curl request. It actually turned out to be that easy, but the access_token that this produced (which was also a valid JWT, leading me to believe I was on the right track) turned out to not be the right thing at all!
+
+Passing it along as a bearer token to the earlier bash script results in a frustrating 301 that redirects you to a 401, giving you no further information as to what, if anything, you did wrong.
+
+Well you're in luck, I'm going to spill the beans on how to avoid making this mistake and wasting an afternoon: Just keep reading the [article](https://learn.microsoft.com/en-us/azure/devops/integrate/get-started/authentication/service-principal-managed-identity?view=azure-devops#q-can-i-use-a-service-principal-or-managed-identity-with-azure-cli), and use the `az account get-access-token` method from the actual example instead of curl'ing the oauth2 endpoint.
+
+If you pass along the accessToken that results from this call and expose it as the AZP_TOKEN to the agent you should roughly see something like this:
+
+{{<img src="/img/blog/self-hosted-agent-startup.png" class="img-fluid" title="Self Hosted Agent starting up" >}}
+
+Great! But I can understand that you don't actually want to clickops your way through this, so I also cooked up a small bit of [terraform](https://dev.azure.com/leebeenen0950/_git/SelfHostedAgent) that should get you up to speed. You'll still have to create at least one App Registration to let Terraform do what it needs to do, refer to the [README](https://dev.azure.com/leebeenen0950/_git/SelfHostedAgent?path=/README.md) for more details.
```

Comment:
clickops → ClickOps

I would rewrite this as: 
"Great! But let's not ClickOps our way through this. I cooked up a bit of Terraform that helps you automate this."

Reply:
Maybe also use some headers (H3 seems to work best in terms of formatting and gaps) to create logical sections. It also draws attention to the footer rather than it mostly disappearing below a screenshot 😉

Reply:
Good recommendation, changed.